nodeprobe/nodeprobe.go

```diff
@@ -2,164 +2,186 @@ package nodeprobe
+// Start func (which can crash the process via logger.Fatal call if it finds out that one of the nodes being
+// probed is not healthy), or it can probe on demand via Probe func.
+// It is used to make sure the Ethereum nodes (CL, EL) are up and running, and they are healthy enough to for
+// SSV node to be able to perform duties.
 type Prober struct {
-	logger           *zap.Logger
-	interval         time.Duration
-	nodes            map[string]Node
-	nodesMu          sync.Mutex
-	healthy          atomic.Bool
-	cond             *sync.Cond
-	unhealthyHandler func()
+	logger *zap.Logger
+
+	// nodesMu protects access to nodes, needed to handle the case when "event syncer" is added dynamically later on
+	// when the Prober is already running.
+	nodesMu sync.Mutex
+	nodes   map[nodeName]Node
 }
 
-func NewProber(logger *zap.Logger, unhealthyHandler func(), nodes map[string]Node) *Prober {
+func NewProber(logger *zap.Logger, consensusClient, executionClient Node) *Prober {
 	return &Prober{
-		logger:           logger,
-		unhealthyHandler: unhealthyHandler,
-		interval:         probeInterval,
-		nodes:            nodes,
-		cond:             sync.NewCond(&sync.Mutex{}),
+		logger: logger,
+		nodes: map[nodeName]Node{
+			el: executionClient,
+
+			// Underlying options.Beacon's value implements nodeprobe.StatusChecker.
+			// However, as it uses spec's specssv.BeaconNode interface, avoiding type assertion requires modifications in spec.
+			// If options.Beacon doesn't implement nodeprobe.StatusChecker due to a mistake, this would panic early.
+			cl: consensusClient,
+		},
 	}
 }
 
 func (p *Prober) Start(ctx context.Context) {
-	go func() {
-		if err := p.Run(ctx); err != nil {
-			p.logger.Error("finished probing nodes", zap.Error(err))
-		}
-	}()
-}
-
-func (p *Prober) Run(ctx context.Context) error {
-	ticker := time.NewTicker(p.interval)
+	ticker := time.NewTicker(probeFrequency)
 	defer ticker.Stop()
 
 	for {
-		p.probe(ctx)
+		func() {
+			probeCtx, cancel := context.WithTimeout(ctx, probeFrequency)
+			defer cancel()
+
+			if err := p.Probe(probeCtx); err != nil {
+				p.logger.Fatal("Ethereum node(s) are either out of sync or down. Ensure the nodes are healthy to resume.", zap.Error(err))
+			}
+		}()
 
 		select {
 		case <-ctx.Done():
-			return ctx.Err()
+			return
 		case <-ticker.C:
 			continue
 		}
 	}
 }
 
-func (p *Prober) probe(ctx context.Context) {
-	// Query all nodes in parallel.
-	ctx, cancel := context.WithTimeout(ctx, p.interval)
+func (p *Prober) Probe(ctx context.Context) error {
+	// Probe all nodes in parallel, use cancel to quit early canceling irrelevant workers.
+	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	var healthy atomic.Bool
-	healthy.Store(true)
 	var wg sync.WaitGroup
+	errsCh := make(chan error)
+
 	p.nodesMu.Lock()
-	for name, node := range p.nodes {
+	nodes := p.nodes
+	p.nodesMu.Unlock()
```

Comment:
The mutex guards only the map pointer, which always remains the same. It doesn't guard the map modification